# ADS 509 Sentiment Assignment

This notebook holds the Sentiment Assignment for Module 6 in ADS 509, Applied Text Mining. Work through this notebook, writing code and answering questions where required. 

In a previous assignment you put together Twitter data and lyrics data on two artists. In this assignment we apply sentiment analysis to those data sets. If, for some reason, you did not complete that previous assignment, data to use for this assignment can be found in the assignment materials section of Blackboard. 


In [193]:
import os
import re
import emoji
import pandas as pd
import numpy as np

from collections import Counter, defaultdict
from string import punctuation

from nltk.corpus import stopwords

sw = stopwords.words("english")

In [195]:
# Add any additional import statements you need here
import warnings
import seaborn

In [197]:
# change `data_location` to the location of the folder on your machine.
data_location = "/Users/clairebentzen/Desktop/MDAS/ADS 509 - Applied Text Mining/Module 2/Assignment2.1/M1 Results/"

# Subfolders
twitter_folder = "twitter/"
lyrics_folder = "lyrics/"

# Specify artist_files for twitter data
artist_files = {'cher':'cher_followers_data.txt',
                'robyn':'robynkonichiwa_followers_data.txt'}

positive_words_file = "positive-words.txt"
negative_words_file = "negative-words.txt"
tidy_text_file = "tidytext_sentiments.txt"

## Data Input

Now read in each of the corpora. For the lyrics data, it may be convenient to store the entire contents of the file to make it easier to inspect the titles individually, as you'll do in the last part of the assignment. In the solution, I stored the lyrics data in a dictionary with two dimensions of keys: artist and song. The value was the file contents. A Pandas data frame would work equally well. 

For the Twitter data, we only need the description field for this assignment. Feel free all the descriptions read it into a data structure. In the solution, I stored the descriptions as a dictionary of lists, with the key being the artist. 




In [200]:
# Read in the lyrics data
# Specify pathway to lyrics folder
lyrics_path = data_location + lyrics_folder

# Create a dataframe to store results
lyrics_data = pd.DataFrame(columns=['artist', 'song', 'lyrics'])

# Iterate through each file in the lyrics folder
for artist in os.listdir(lyrics_path):
    artist_path = os.path.join(lyrics_path, artist)
    
    # Iterate through each file in the artist folders
    for song in os.listdir(artist_path):
        song_path = os.path.join(artist_path, song)
        rem_prefix = song.removeprefix(f'{artist}_')
        song_title = rem_prefix.removesuffix('.txt')

        # Open and read the contents of the file (song)
        with open(song_path, 'r') as file:
            contents = file.read()
            # Prepare data to add to dataframe
            data = {'artist': artist, 'song': song_title, 'lyrics': contents}
            # The df.append() function is deprecated, so we will ignore warnings here
            with warnings.catch_warnings():
                warnings.simplefilter('ignore')
                # Append row of data to lyrics_df
                lyrics_data = lyrics_data.append(data, ignore_index=True)

In [202]:
# View lyrics_data dataframe
lyrics_data.head()

artist          song                                             lyrics
0  robyn  includemeout  "Include Me Out"\n\n\n\nIt is really very simp...
1  robyn      electric  "Electric"\n\n\n\nElectric...\n\nIt's electric...
2  robyn     beach2k20  "Beach 2K20"\n\n\n\n(So you wanna go out?\nHow...
3  robyn     lovekills  "Love Kills"\n\n\n\nIf you're looking for love...
4  robyn   timemachine  "Time Machine"\n\n\n\nHey, what did I do?\nCan...

In [204]:
# Read in the twitter data
# Read cher twitter data
twitter_data = pd.read_csv(data_location + twitter_folder + artist_files['cher'],
                           sep="\t",
                           quoting=3)

twitter_data['artist'] = "cher"

# Read robyn twitter data
twitter_data_2 = pd.read_csv(data_location + twitter_folder + artist_files['robyn'],
                             sep="\t",
                             quoting=3)
twitter_data_2['artist'] = "robyn"

# Concat twitter dataframes
twitter_data = pd.concat([
    twitter_data,twitter_data_2])
    
del(twitter_data_2)

In [205]:
# Keep only artist and description columns
twitter_data = twitter_data[['description', 'artist']]

In [163]:
# Read in the positive and negative words and the
# tidytext sentiment. Store these so that the positive
# words are associated with a score of +1 and negative words
# are associated with a score of -1. You can use a dataframe or a 
# dictionary for this.

# Read positive words and add score column
pos_words = pd.read_csv('positive-words.txt', delimiter='\t', skiprows=35, header=None)
pos_words['score'] = 1
pos_words.columns.values[0] = 'word'

# Read negative words and add score column
neg_words = pd.read_csv('negative-words.txt', delimiter='\t', skiprows=35, header=None)
neg_words['score'] = -1
neg_words.columns.values[0] = 'word'

# Read tidytext sentiment
tidytext = pd.read_csv('tidytext_sentiments.txt', delimiter='\t')
tidytext.drop(columns='lexicon', inplace=True)
# Replace sentiment with -1 and 1
tidytext['score'] = tidytext['sentiment'].apply(lambda x: -1 if x == 'negative' else 1)
tidytext.drop(columns='sentiment', inplace=True)

# Combine dataframes
words = pd.concat([pos_words, neg_words, tidytext])
words

word  score
0             a+      1
1         abound      1
2        abounds      1
3      abundance      1
4       abundant      1
...          ...    ...
15128        win      1
15129     winner      1
15130    winners      1
15131    winning      1
15132     worthy      1

[21922 rows x 2 columns]

In [208]:
# View twitter_data dataframe
twitter_data.head()

description artist
0                                                NaN   cher
1           𝙿𝚛𝚘𝚞𝚍 𝚜𝚞𝚙𝚙𝚘𝚛𝚝𝚎𝚛 𝚘𝚏 𝚖𝚎𝚜𝚜𝚢 𝚋𝚞𝚗𝚜 & 𝚕𝚎𝚐𝚐𝚒𝚗𝚐𝚜   cher
2          163㎝／愛かっぷ💜26歳🍒 工〇好きな女の子💓 フォローしてくれたらDMします🧡   cher
3                                                csu   cher
4  Writer @Washinformer @SpelmanCollege alumna #D...   cher

## Sentiment Analysis on Songs

In this section, score the sentiment for all the songs for both artists in your data set. Score the sentiment by manually calculating the sentiment using the combined lexicons provided in this repository. 

After you have calculated these sentiments, answer the questions at the end of this section.


In [210]:
# Punctuation
punctuation = set(punctuation) # speeds up comparison

# Stopwords
sw = stopwords.words("english")

# Removes stopwords
def remove_stop(tokens):
    
    tokens = [word for word in tokens if word not in sw]
    
    return(tokens)
 
# Removes punctuation
def remove_punctuation(text, punct_set=punctuation): 
    
    return("".join([ch for ch in text if ch not in punct_set]))

# Tokenizes the text
def tokenize(text):     
    
    return text.split()

# Create function to calculate sentiment score
def calculate_score(text): 

    # Lower case
    text = str.lower(text)

    # Remove punctuation
    text = remove_punctuation(text)

    # Tokenize text
    tokens = tokenize(text)

    # Remove stopwords
    tokens = remove_stop(tokens)

    # sentiment_score will contain the total of sentiments for words found in the lexicon per song
    sentiment_score = 0
    # counter will contain the number of words that were actually found in the lexicon
    counter = 0

    # Add sentiments for words found in lexicon
    for word in tokens:
        if word in words['word'].values:
            counter += 1
            score = words.loc[words['word'] == word, 'score'].values[0]
            sentiment_score += score

    # Return average sentiment of the song, and return 0 if no words were found in the lexicon
    try:        
        return sentiment_score / counter
    except ZeroDivisionError:
        return 0

In [212]:
# Calculate sentiment scores for lyrics_data
lyrics_data['sentiment_score'] = lyrics_data['lyrics'].apply(calculate_score)

In [213]:
# Find average sentiment score by artist
lyrics_avg = lyrics_data.groupby('artist')['sentiment_score'].mean()
lyrics_avg

artist
cher     0.181567
robyn    0.282513
Name: sentiment_score, dtype: float64

Q: Overall, which artist has the higher average sentiment per song? 

A: Overall, Robyn has the highest average sentiment per song. The average for Robyn is 0.283 and the average for Cher is 0.182.

In [238]:
pd.set_option('display.max_colwidth', None)

# Subset cher lyrics
cher_lyrics = lyrics_data[lyrics_data['artist'] == 'cher']

# Get the top 3 scores
cher_top_3 = cher_lyrics.nlargest(3, 'sentiment_score')
cher_top_3

artist                song  \
119   cher              mylove   
407   cher  youwouldntknowlove   
355   cher        mylove318663   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        lyrics  \
119                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   "My Love"\n\n\n\nWhen I go away\nI know my heart can stay with my love\nIt's understood\nEverywhere with my love\nMy love does it good, whoa\nMy love, oh only my love\nMy love does it good\n\nAnd when the cupboard's bare\nI'll still find something there with my love\nIt's understood\nEverywhere with my love\nMy love does it so good, whoa\nMy love, oh only my love\nMy love does it good\n\nOnly my love, oh only my love\nOnly my love, hold the other things to me\nOh, only my love, oh only my love\nOnly my love does it good to me\n\nEverywhere with my love\nDon't you see my love\nIt does it so good\nOh, only my love, only my love\nMy love does it good\n\nDon't ever ask me why\nI never say "good-bye" to my love\nIt's understood\n'Cause everywhere with my love\nMy love does it good, whoa\nOnly my love, oh only my love\nMy love does it good\n\nOh only my love\nOnly my love does it good to me,whoa\n   
407  "You Wouldn't Know

In [240]:
# Get the bottom 3 scores
cher_bottom_3 = cher_lyrics.nsmallest(3, 'sentiment_score')
cher_bottom_3

artist                      song  \
219   cher                  skindeep   
324   cher  chiquititaspanishversion   
344   cher          crymyselftosleep   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           lyrics  \
219  "Skin Deep"\n\n\n\nYou came on to me a house of fire\nOne look was all it took for my desire\nI'm still burnin'\nIt's the middle of the night\nI need somethin'\nThat can make me feel alright as I go\n\nSkin deep, to the bone\nEvery time I see your face\nSkin deep, to the bone\nTellin' me that I'm in trouble\nS.O.S.\nI'm in distress, tonight\n\nKnock knock in the dark, what will I find?\nYour eyes, hypnotize but can they read my mind\nWatch your step now\nWe're gettin' closer to the edge\nThis is dangerous\nAnd all the lights are flashin' red as I go\n\nSkin deep, to the bone\nEvery time I see your face I go\nSkin deep, to the bone\nTellin' me that I'm in trouble\n\nSkin deep, to the bone\nEvery time I hear your voice I go\nSkin deep, to the bone\nTellin' me that I'm in trouble\nToo far gone\nI can't hold on, tonight, not tonight\n\nSkin deep, skin deep\nI'm still burnin'\nIt's the middle of the night\nI need someone\nWho can make me feel alright as I go\n\nSkin deep, to the bone\nEvery time I see your face I go\nSkin deep, to the bone\nTellin' me that I'm in trouble\n\nSkin deep, to the bone\nEvery time I see your face I go\nSkin deep, to the bone\nTellin' me that I'm in trouble\n\nSkin deep, to the bone\nEvery time I hear your voice I go\nSkin deep, to the bone\nTellin' me that I'm in trouble\nSkin deep, to the bone\n   
324                                                                                            "Chiquitita (Spanish Version)"\n\n\n\nChiquitita, dime por qué\nTu dolor hoy te encadena\nEn tus ojos hay, una sombra de gran pena\nNo quisiera verte así\nAunque quieras disimularlo\nSi es que tan triste estás para qué quieres callarlo\n\nChiquitita, dímelo tu\nEn mi hombro, aquí llorando\nCuenta conmigo ya, para así seguir andando\nTan segura te conocí\nY ahora tu ala quebrada\nDéjamela arreglar, yo la quiero ver curada\n\nChiquitita, sabes muy bien\nQue las penas vienen y van y desaparecen\nOtra vez vas a bailar\nY serás feliz\nComo flores que florecen\nChiquitita, no hay que llorar\nLas estrellas brillan por ti allá en lo alto\nQuiero verte sonreír\nPara compartir\nTu alegría, Chiquitita\nOtra vez quiero compartir\nTu alegría, Chiquitita\n\nChiquitita, no hay que llorar\nQuiero verte sonreír\nPara compartir tu alegría\nHey, Chiquitita\n\nChiquitita, sabes muy bien\nQue las penas vienen y van y desaparecen\nOtra vez vas a bailar\nY serás feliz\nComo flores que florecen\nChiquitita, no hay que llorar\nLas estrellas b

Q: For your first artist, what are the three songs that have the highest and lowest sentiments? Print the lyrics of those songs to the screen. What do you think is driving the sentiment score? 

A: Cher's top 3 and bottom 3 songs based on sentiment are printed above. For the positive sentiment songs (My Love, You Wouldn't Know Love, and My Love 318663), the lyrics include the word love repeatedly which probably drives the positive sentiment. For the negative sentiment songs (Skin Deep, Chiquititas Spanish version, and Cry Myself to Sleep), the lyrics include words such as fire, trouble, cry, dangerous, distress, and dark which drives the negative score. I think Chiquititas is mistakenly included in this list because it is in Spanish, so the sentiment is not accurately calculated.

In [236]:
# Subset robyn lyrics
robyn_lyrics = lyrics_data[lyrics_data['artist'] == 'robyn']

# Get the top 3 scores
robyn_top_3 = robyn_lyrics.nlargest(3, 'sentiment_score')
robyn_top_3

artist                song  \
60  robyn  jagvetendejligrosa   
64  robyn       babyforgiveme   
67  robyn            monument   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   lyrics  \
60                                                                                                                                                                                                                                                "Jag Vet En Dejlig Rosa"\n\n\n\nJag vet en dejlig rosa och vit som liljeblad\nNär jag på henne tänker så görs mitt hjärta glad\nDess stämma ger en hjärtans tröst\nLik näktergalens blida röst\nSå fager och så ljuv\n\nSom solen fagert skiner är hon som purpur klar\nGud låt dig aldrig sörja men alltid vara glad\nMå de få komma samman\nMed hjärtans fröjd och gamman\nSom längta till varann\n\nVar dag går solen neder och dagelig uppgår\nNär kommer dagen blider att jag dig skåda får\nI hågen är du jämt mig när\nFarväl, farväl min hjärtans kär\nMångtusende godnatt\n   
64                                                                                                  "Baby Forgive Me"\n\n\n\nHere come the night\nIn your eyes\nBaby, be brave\nBe wise\nIt's up to you\nYou do what you like, mmm\nWon't you give it a chance, baby?\nJust one more try\n\nBaby forgive me\nBaby forgive me\nBaby forgive me\nBaby forgive me\n\n(Won't you give it a chance, baby?)\nWon't you give it a chance, baby?\nJust one more try\nOne more try\nYeah, eh\n\nYou got the power\nYou set the price\nBut baby, be fair\nBe nice\nYou say you want to be happy\nThen you got to put your love on the line\nJust let me make you smile again, baby\nI know we can work it out\nYes, I know we can\n\nBaby forgive me\nYeah\nBaby forgive me\nBaby forgive me\nBaby forgive me\nBaby forgive me\nBaby forgive me\nBaby forgive me\nBaby forgive me\n   
67  "Monument"\n\n\n\nMake a space\nFor my body\nDig a hole\nPush the sides apart\nThis is what\nI'm controlling\nIt's a moat\nThe inside that I carve\n\nThis will my monument\nThis will be a beacon when I'm gone\nGone, gone\nWhen I'm gone\nGone, gone\nWhen I'm gone\n\nSo that when the moment comes\nI can say I did it all with love\nLove, love\nAll with love\nLove, love\nAll with love\n\nMake a cast\nOf my body\nPull back out\nSo that I can see\nLet go of\nHow you knew me\nLet go of\nWhat I used to be\n\nI will let this monument\nRepresent a moment of my life\nLife, life\nOf my life\nLife, life\nOf my life\n\nI will let this monument\nRepresent a moment of my life\nLife, life\nOf my life\nLife, life\nOf my life\n\nMake a cast\nOf my body\nPull back out\nSo that I can see\nLet go of\nHow you knew me\nLet go of\nWhat I used to be\n   

    sentiment_score  
60              1.0  
64              1.0  
67              1.0

In [242]:
# Get the bottom 3 scores
robyn_bottom_3 = robyn_lyrics.nsmallest(4, 'sentiment_score')
robyn_bottom_3

artist                             song  \
53  robyn        dontfuckingtellmewhattodo   
75  robyn  dontfuckingtellmewhattodo114520   
16  robyn                   criminalintent   
95  robyn                 crashandburngirl   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               lyrics  \
53                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            "Don't Fucking Tell Me What To Do"\n\n\n\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\nMy drinking is killing me\n\nMy smoking is killing me\nMy diet is killing me\nMy heels are killing me\nMy shoppin

Q: For your second artist, what are the three songs that have the highest and lowest sentiments? Print the lyrics of those songs to the screen. What do you think is driving the sentiment score? 

A: Robyn's top 3 and bottom 3 songs based on sentiment are printed above. For the positive sentiment songs (Jag Vet En Dejlig Rosa, Baby Forgive Me, and Monument), the lyrics include the words power, love, forgive, monument, and smile which probably drives the positive sentiment. I think Jag Vet En Dejlig Rosa is mistakenly included in this list because it is in Spanish, so the sentiment is not accurately calculated. For the negative sentiment songs (Don't Fucking Tell Me What To Do, Criminal Intent, and Crash and Burn Girl), the lyrics include words such as killing, criminal, hurt, crash, and burn which drives the negative score.

Q: Plot the distributions of the sentiment scores for both artists. You can use `seaborn` to plot densities or plot histograms in matplotlib.

Q: Plot the distributions of the sentiment scores for both artists. You can use `seaborn` to plot densities or plot histograms in matplotlib.

In [135]:
# your code here
data = {'emoji': ['😂', '❤️', '😭', '😍', '✨', '♥️', '😊', '😔', '💕', '😘', 
                  '😒', '😩', '😁', '🔥', '💔', '😢', '😳', '😡', '😞', '😑', '🙃'],
        'score': [1, 1, -1, 1, 1, 1, 1, -1, 1, 1, 
                  -1, -1, 1, 1, -1, -1, -1, -1, -1, -1, -1]}
emojis = pd.DataFrame(data)

## Sentiment Analysis on Twitter Descriptions

In this section, define two sets of emojis you designate as positive and negative. Make sure to have at least 10 emojis per set. You can learn about the most popular emojis on Twitter at [the emojitracker](https://emojitracker.com/). 

Associate your positive emojis with a score of +1, negative with -1. Score the average sentiment of your two artists based on the Twitter descriptions of their followers. The average sentiment can just be the total score divided by number of followers. You do not need to calculate sentiment on non-emoji content for this section.

In [137]:
# Checks if contains emojis
def contains_emoji(s):
    
    s = str(s)
    emojis = [ch for ch in s if emoji.is_emoji(ch)]

    return(len(emojis) > 0)

# Check which twitter descriptions have emojis
twitter_data['has_emoji'] = twitter_data["description"].apply(contains_emoji)

In [138]:
twitter_data = twitter_data[twitter_data['has_emoji']==True]

In [141]:
# Create function to calculate sentiment score with emojis
def calculate_score_emojis(text): 

    # Lower case
    text = str.lower(text)

    # Remove punctuation
    text = remove_punctuation(text)

    # Tokenize text
    tokens = tokenize(text)

    # Remove stopwords
    tokens = remove_stop(tokens)

    # 
    sentiment_score = 0
    
    for word in tokens:
        if word in emojis['emoji'].values:
            score = emojis.loc[emojis['emoji'] == word, 'score'].values[0]
            sentiment_score += score
            
    return sentiment_score / len(tokens)

Q: What is the average sentiment of your two artists? 

A: <!-- Your answer here --> 

---

Q: Which positive emoji is the most popular for each artist? Which negative emoji? 

A: <!-- Your answer here --> 



In [143]:
#
twitter_data['sentiment_score'] = twitter_data['description'].apply(calculate_score_emojis)
twitter_data

In [147]:
twitter_avg = twitter_data.groupby('artist')['sentiment_score'].mean()
twitter_avg

artist
cher     0.014761
robyn    0.013571
Name: sentiment_score, dtype: float64